### PR TITLE
New extract_ipv6 algorithm

### PR DESCRIPTION
**Pull request by dkerr64**
New algorithm for `extract_ipv6` using only regex and based on that documented in the _Regular Expressions Cookbook by J.Goyvaerets and S.Levithan_ published by O'Reilly (https://github.com/shihyu/Regular_Expressions/blob/master/O'Reilly-Regular%20Expressions%20Cookbook(2nd%20Edition).pdf)

### CODE REVIEW

**Comment by rhansen** [comment]
We need to make sure we don't violate any copyright licenses. The vagueness of the "Using Code Examples" section on page xiii worries me.

**Comment by SuperSandro2000** [comment]
One Regex is not a significant portion of the book.
But I would like to document that in code so that we easier know from where we took the regex and who to blame if it does not work 😛

**Comment by dkerr64** [comment]
I think the U_sing Code Examples_ section is quite clear that it is ok to use code samples.  I have no problem inserting attribution into the comments in the code and will do that.

**Comment by dkerr64** [comment]
> Once PR #161 is merged would you rebase this and add some tests? See PR #175 for an example you can follow.

Yes I will work on that.

**Comment by dkerr64** [comment]
@rhansen I need your help with unit tests.  I have created a test file and added it to the makefile, but when I run I get the following...
```
$ make check
make  check-TESTS
make[1]: Entering directory '/home/david/github/ddclient'
make[2]: Entering directory '/home/david/github/ddclient'
ERROR: t/is-and-extract-ipv6.pl - missing test plan
ERROR: t/is-and-extract-ipv6.pl - exited with status 255 (terminated by signal 127?)
PASS: t/version.pl 1 - version matches Autoconf config
PASS: t/version.pl 2 - no (unexpected) warnings (via done_testing)
============================================================================
Testsuite summary for ddclient 3.9.1
============================================================================
# TOTAL: 4
# PASS:  2
# SKIP:  0
# XFAIL: 0
# FAIL:  0
# XPASS: 0
# ERROR: 2
============================================================================
See ./test-suite.log
============================================================================
make[2]: *** [Makefile:591: test-suite.log] Error 1
make[2]: Leaving directory '/home/david/github/ddclient'
make[1]: *** [Makefile:699: check-TESTS] Error 2
make[1]: Leaving directory '/home/david/github/ddclient'
make: *** [Makefile:891: check-am] Error 2
```
I don't know what I am doing wrong.  Can you have a look please.

Thank you.

**Comment by rhansen** [comment]
In your `t/is-and-extract-ipv6.pl` file, replace `is_ipv6` with `ddclient::is_ipv6`. (The `ddclient` script is imported as a module, and all of its functions are in the `ddclient` namespace.)

**Comment by rhansen** [comment]
Tip: Enable verbose mode to see the errors without having to look at the `.log` file: `make VERBOSE=1 check`

**Comment by dkerr64** [comment]
> In your `t/is-and-extract-ipv6.pl` file, replace `is_ipv6` with `ddclient::is_ipv6`. (The `ddclient` script is imported as a module, and all of its functions are in the `ddclient` namespace.)

Thank you, now I can build out my unit tests.

**Comment by dkerr64** [comment]
@rhansen this one is ready now (pending your review).  Any enhancements to the unit test I will do as part of PR #197
This was a good learning experience (adding tests, regex to die for).

**Comment by SuperSandro2000** [comment]
I like that you didn't merge master into this one. Not sure if all the commits are necessary but I don't want to nit pick. Good work! 😃

**Comment by SuperSandro2000** [comment]
I really like that the tests cover so many cases and also test some ip command extractions. If we fix some bugs related to this later we should expand them but this will probably be done word edge cases.

**Comment by dkerr64** [comment]
> I like that you didn't merge master into this one. Not sure if all the commits are necessary but I don't want to nit pick. Good work! 😃

yup, I should have squashed everything together.  Still learning the intricacies of git.

File: t/is-and-extract-ipv6.pl [hunk 1]
```diff
@@ -0,0 +1,531 @@
...
+    "3001:DB8:4341:0781:1111:2222:3333:4444",
+    "3001:DB8:4341:0781::4444",
+    "3001:DB8:4341:0781:1111::",
+    "3001:DB8:4341:0781::100",
+    "3001:DB8:4341:0781::1",
+    "3001:DB8:4341:0781::0001",
+);
+
+# For ULA addresses we randomly generated a /48 prefix per RFC 4193
+# and are using subnet ID of 1 because why not.
```
Comment (rhansen) [inline t/is-and-extract-ipv6.pl:26]: Include the first and last of the ULA block (fc00:: and fdff:ffff:ffff:ffff:ffff:ffff:ffff:ffff) in this list.

File: t/is-and-extract-ipv6.pl [hunk 2]
```diff
@@ -0,0 +1,531 @@
...
+    "fdb6:1d86:d9bd:1::1",
+    "fdb6:1d86:d9bd:1::0001",
+);
+
+# 
+my @valid_ipv6_lla = (
+    "fe80::1111:2222:3333:4444",
+    "fe80::4444",
+    "fe80::1111",
+);
```
Comment (rhansen) [inline t/is-and-extract-ipv6.pl:41]: Include the first and last addresses of the block (fe80:: and febf:ffff:ffff:ffff:ffff:ffff:ffff:ffff) in this list.

File: t/is-and-extract-ipv6.pl [hunk 3]
```diff
@@ -0,0 +1,531 @@
...
+
+
+#######################################################################
+## Run through a bunch of IPv6 addresses
+foreach my $ip (@all_valid_ipv6) {
+    is(ddclient::is_ipv6($ip),1,"Testing valid 'is_ipv6($ip)'");
+}
+
+# Run through valid mixed (IPv4 within IPv6 address).  We do not accept
+# this format so expect valid check to fail.
```
Comment (rhansen) [inline t/is-and-extract-ipv6.pl:467]: Why don't we? Short-term implementation convenience? If so, mark them as TODO.

Comment (dkerr64) [reply t/is-and-extract-ipv6.pl:467]: Permanent unless there is a need to set a mixed IPv6/IPv4 address to a DNS server.  I'm not even sure that it is possible.  Is it legal to set something like `2001:DB8::123.123.123.123` to a DNS server?  I've never seen that.  If it is possible then we can decide to support it, but for now it was deliberate choice not to.

File: t/is-and-extract-ipv6.pl [hunk 4]
```diff
@@ -0,0 +1,531 @@
+use Test::More;
+
+SKIP: { eval { require Test::Warnings; } or skip($@, 1); }
+eval { require 'ddclient'; } or BAIL_OUT($@);
+
+# Note, for GUA addresses we use 2001:DB8::/32 as that is reserved for
+# documentation per RFC 3849 and so not routable on public internet.
+# But we do have to add some others starting in 3xxx:: because they 
+# are also valid GUA.
```
[the remark this answers is not in the document]
Comment (dkerr64) [reply t/is-and-extract-ipv6.pl:9]: I will add the lower and upper limits.  I just documented that I chose non-routable IPv6 prefixes to deter a malicious actor.

File: ddclient.in [hunk 5]
```diff
@@ -2276,44 +2276,37 @@ sub extract_ipv4 {
     return $ip;
 }
 
+######################################################################
+## Regex to find IPv6 in standard or compressed format. Mixed IPv6/IPv4
+## not supported. Accepts leading zeros in the address. Adapted from
+## regex in Regular Expressions Cookbook by J.Goyvaerets and S.Levithan
+######################################################################
+my $regex_ipv6 = "(?<![:.0-9A-Z])              ## Negative lookbehind for word boundry
```
[the remark this answers is not in the document]
Comment (dkerr64) [reply ddclient.in:2284]: Can't.  The regex itself has the boundary checks (lookaheads) embedded in the middle of the regex not just at the start and end.  In other words they are needed for the `is_ipv6` case.  If all we had to do was test for rather than extract as well then we could use a simpler boundry check, but then we would have two different regex's.  I think it is better to have just one.

File: ddclient.in [hunk 6]
```diff
@@ -2276,44 +2276,37 @@ sub extract_ipv4 {
...
 ######################################################################
 ## is_ipv6() validates if string is valid IPv6 address with no preceding
-## or trailing spaces/characters.
+## or trailing spaces/characters, not even line breaks.
 ######################################################################
 sub is_ipv6 {
-    my ($value) = @_;
-    return (length($value // '') != 0) &&
-        ((extract_ipv6($value) // '') eq (($value =~ s/\b0+\B//g) ? $value : $value));
+    return ((shift // '') =~ /\A$regex_ipv6\z/xi) ne '';
```
[the remark this answers is not in the document]
Comment (dkerr64) [reply ddclient.in:2299]: This was one of the unexpected findings from creating unit tests. We need to specifically test for start of string and end of string `\A` and `\z` not for start and end of line.  If we only test for line starts then the regex fails to work when a valid IP address is wrapped by blank lines (returns valid when it should return invalid).

Prior implementations of is_ipv6 returned `1`/true or `''`.  Without` ne ''` this function returned the value passed in.  With it we get same as prior implementations.

File: ddclient.in [hunk 7]
```diff
@@ -2276,44 +2276,37 @@ sub extract_ipv4 {
...
+######################################################################
+my $regex_ipv6 = "(?<![:.0-9A-Z])              ## Negative lookbehind for word boundry
+    ( (?:[0-9A-F]{1,4}:){7}[0-9A-F]{1,4}       ## Is in standard format
+     |(?=(?:[0-9A-F]{0,4}:){0,7}[0-9A-F]{0,4}  ## OR compressed with at most 7 colons
+      (?![:.0-9A-Z]))                          ##   and negative lookahead for boundry
+      (?:(?:[0-9A-F]{1,4}:){1,7}|:)(?:(?::[0-9A-F]{1,4}){1,7}|:)  ## and only 1 double-colon
+     |(?:[0-9A-F]{1,4}:){7}:                   ## OR compressed with 8 colons (double at end)
+     |:(?::[0-9A-F]{1,4}){7}                   ## OR compressed with 8 colons (double at front)
+    )                                          ## End of first capture group
+    (?![:.0-9A-Z])";                           ## Negative lookahead for boundry
```
[the remark this answers is not in the document]
Comment (dkerr64) [reply ddclient.in:2292]: I was having problems with that during unit testing.  But looks like it might have been something else going on as when I test it now it is working.  Thanks for prompting me to look at it again.  I will update the code.

File: ddclient.in (hunk 6 above)
[the remark this answers is not in the document]
Comment (dkerr64) [reply ddclient.in:2299]: Further testing and yes we need `ne ''` in order to return a `1` or `''`.

**Comment by dkerr64** [comment]
> Additional cases to test:
> 
> * `extract_ipv6` on `''`, `undef`, and a string that doesn't contain an IPv6 address
> * `is_ipv6` on `''` and `undef`
> * `extract_ipv6` on an address with a zone index
> * `is_ipv6` on an address with a zone index

Empty string was already in my list of invalid IPv6's.  I am adding additional samples of valid IPv6 addresses that have text before and/or after that would make them invalid.

I will add explicit test for undef inputs.

File: t/is-and-extract-ipv6.pl (hunk 1 above)
Comment (dkerr64) [reply t/is-and-extract-ipv6.pl:26]: Added.  But note that valid ULAs must all be in the range fd::/7.  While RFC 4193 documents fc::/7 the last bit must always be set to 1 to indicate locally defined.  Set to 0 is not defined and reserved for future use, therefore all ULAs in use today must be fd::/7.

File: t/is-and-extract-ipv6.pl (hunk 2 above)
Comment (dkerr64) [reply t/is-and-extract-ipv6.pl:41]: Added

File: t/is-and-extract-ipv6.pl (hunk 3 above)
Comment (dkerr64) [reply t/is-and-extract-ipv6.pl:494]: Okay, so I decided to try and set an AAAA record at cloudflare for a valid address inside my network `2601:xxxx:yyyy:e723::192.168.17.2` and see what happens.  Cloudflare accepted it and when I do a DNS request I get given the address `2601:xxxx:yyyy:e723::c0a8:1102` and pings to that appear to get routed correctly to my gateway... but of course don't get through because of my firewall.

Which means, I think we should support it.  But can we open it as an issue rather than hold up this PR please.

File: ddclient.in (hunk 6 above)
[the remark this answers is not in the document]
Comment (rhansen) [reply ddclient.in]: > We need to specifically test for start of string and end of string `\A` and `\z` not for start and end of line.

By default `^` and `$` only match the beginning and end of the string. Only if [the `/m` modifier](https://perldoc.perl.org/5.10.1/perlre.html#Modifiers) is used will they also match start and end of line.

> we need `ne ''` in order to return a `1` or `''`.

I think it's OK to return any truthy or falsy value. If we want to restrict the return values to `1` and `''`, then I would rather do `&& 1 || ''`.

File: t/is-and-extract-ipv6.pl (hunk 3 above)
Comment (rhansen) [reply t/is-and-extract-ipv6.pl:494]: > I think we should support it. But can we open it as an issue rather than hold up this PR please.

Sure. Just mark those as TODO:

```perl
for my $ip (@valid_mixed_ipv6) {
    TODO: {
        local $todo = "IPv6 addresses with embedded IPv4 addresses not yet supported";
        ok(ddclient::is_ipv6($ip), 1, "is_ipv6('$ip')");
    }
}
```

**Comment by rhansen** [comment]
> We'll never see it when using `ip` command as we will always use scope global.

True, but we might see it for `cmdv6`.

File: ddclient.in (hunk 6 above)
[the remark this answers is not in the document]
Comment (dkerr64) [reply ddclient.in]: Well if I use `$` the unit test fails.  I would have to remove the test for an IP address wrapped in `\n`.  Here is how it is explained at https://perldoc.perl.org/perlre.html

_The `\A` and `\Z` are just like `"^"` and `"$"` , except that they won't match multiple times when the `/m` modifier is used, while `"^"` and `"$"` will match at every internal line boundary. To match the actual end of the string and not ignore an optional trailing newline, use `\z` ._  (that is lowercase `\z `which is significant).

File: t/is-and-extract-ipv6.pl (hunk 3 above)
Comment (dkerr64) [reply t/is-and-extract-ipv6.pl:494]: Give me a day to look at it.  I think it might be fairly easy to add the support now.

**Comment by dkerr64** [comment]
> True, but we might see it for `cmdv6`. Those should be accepted.

Yes we should expect to see that and it will be accepted because reply from `cmdv6` will go through `extract_ipv6` which will find the IP address leaving out the zone index.  So `is_ipv6` should reject it, `extract_ipv6` should work with it.  And that is the testing that we should do.

File: t/is-and-extract-ipv6.pl (hunk 3 above)
Comment (dkerr64) [reply t/is-and-extract-ipv6.pl]: Mixed IPv6/IPv4 format now supported.  Ready for another review @rhansen

File: ddclient.in [hunk 8]
```diff
@@ -2273,65 +2273,79 @@ sub get_ip {
...
+######################################################################
+## Regex to find IPv6 in standard, compressed format. Accepts
+## embedded leading zeros in the address. Adapted from regex in the
+## Regular Expressions Cookbook by J.Goyvaerets and S.Levithan
+######################################################################
+my $regex_ipv6_only = qr/(?<![:.0-9A-Z])        ## Negative lookbehind for word boundry
+    (?:[0-9A-F]{1,4}:){7}[0-9A-F]{1,4}          ## Is in standard format
+      |(?=(?:[0-9A-F]{0,4}:){0,7}[0-9A-F]{0,4}  ## OR compressed with at most 7 colons
+       (?![:.0-9A-Z]))                          ##   and negative lookahead for boundry
+       (?:(?:[0-9A-F]{1,4}:){1,7}|:)(?:(?::[0-9A-F]{1,4}){1,7}|:)  ## and only 1 double-colon
```
Comment (rhansen) [inline ddclient.in:2291]: This case in particular is really difficult to read. I know it's correct, but I think we can write something that's easier to understand. What do you think about this:

```perl
my $regex_ipv6_only = qr/
  # Define some named groups so we can use Perl's recursive subpattern feature for shorthand:
  (?<g>[0-9A-F]{1,4}){0}  # "g" matches a group of 1 to 4 hex chars
  (?<g_>(?&g):){0}        # "g_" matches a group of 1 to 4 hex chars followed by a colon
  (?<_g>:(?&g)){0}        # "_g" matches a colon followed by a group of 1 to 4 hex chars
  (?<g0>(?&g)?){0}        # "g0" is an optional "g" (matches a group of 0 to 4 hex chars)
  (?<g0_>(?&g0):){0}      # "g0_" is an optional "g" followed by a colon
  (?<x>[:.0-9A-Z]){0}     # "x" matches chars that should never come before or after the address

  # Now for the actual regular expression:

  # Not preceded by a character that is not allowed to come before the address
  (?<!(?&x))
  (?:
    # Exactly 8 groups of 1-4 hex chars
    (?&g_){7}(?&g)
    # OR compressed form with the double colon at the end
    | (?&g_){1,7}:
    # OR compressed form with the double colon at the beginning
    | :(?&_g){1,7}
    # OR compressed form with the double colon at both ends :)
    | ::
    # OR compressed form with the double colon in the middle
    | (?=                   # Only consider this case if the string...
        (?&g0_){2,7}(?&g0)  # ...has 3 to 8 possibly empty groups (at least 3 because there
                            # will be at least one group before the ::, at least one group
                            # after the ::, and the :: itself is an empty group)...
        (?!(?&x))           # ...then ends.
      )
      (?&g_){0,6}(?&g)  # 1 to 7 non-empty groups before the double colon
      ::
      (?&g_){0,6}(?&g)  # 1 to 7 non-empty groups after the double colon
  )
  # Not followed by a character that is not allowed to come after the address
  (?!(?&x))
/xi;
```

File: ddclient.in [hunk 9]
```diff
@@ -2273,65 +2273,79 @@ sub get_ip {
...
+## Regex to find IPv4 address. Accepts embedded leading zeros.
+######################################################################
+my $regex_ipv4 = qr/(?:(?<octet>25[0-5]|2[0-4]\d|[01]?\d\d?)\.){3}(?&octet)/;
+
+######################################################################
+## Regex to find IPv6 in standard, compressed format. Accepts
+## embedded leading zeros in the address. Adapted from regex in the
+## Regular Expressions Cookbook by J.Goyvaerets and S.Levithan
+######################################################################
+my $regex_ipv6_only = qr/(?<![:.0-9A-Z])        ## Negative lookbehind for word boundry
```
Comment (rhansen) [inline ddclient.in:2287]: I think everything after the negative lookbehind here and before the negative lookahead at the end must be inside a group: `(?:...)`.

File: ddclient.in (hunk 8 above)
Comment (dkerr64) [reply ddclient.in:2291]: Thats an interesting idea.  Let me run it through my tests.

Comment (dkerr64) [reply ddclient.in:2291]: So that passes all my unit tests.  But interestingly Regex101 thinks there is an error in the pattern [(see here)](https://regex101.com/r/uG5k0D/1) but debuggex is quite [happy](https://www.debuggex.com/r/TFvjD9Kf8iT64BOK) with it.  Looks like you broke Regex101 !!

Comment (rhansen) [reply ddclient.in:2291]: One possible downside to my suggested regex is this bit from [perl5200delta](https://perldoc.perl.org/perl5200delta.html#Regular-Expressions) (changelog for Perl v5.20):

> Under certain conditions, Perl would throw an error if in a lookbehind assertion in a regexp, the assertion referred to a named subpattern, complaining the lookbehind was variable when it wasn't. This has been fixed. [perl #120600], [perl #120618]. The current fix may be improved on in the future.

This sounds like the Regex101 error. I don't think we need to do anything unless some of the CI runs fail. If they do, we'll have to replace `(?<!(?&x))` with `(?<![:.0-9A-Z])`.

Comment (dkerr64) [reply ddclient.in:2291]: So you got me worried there because MacOS has 5.18.  So I just ran this through my tests on my Mac and it passed.  But we probably need to test on older versions of MacOS / OS X.  I'm looking at my other regex (the one that handles mixed IPv6/IPv4) to adopt your idea into that.  I've tested pulling out the named group definitions into their own `qr/.../` and then referencing that in the main regex's and it works 👍

Comment (dkerr64) [reply ddclient.in]: @rhansen You totally inspired me.  Check out the last commit.

**Comment by rhansen** [comment]
Looking nice. I think we can combine the separate regexes into a single variable without losing readability. I pushed a commit to your branch that does just that; please take a look and tell me what you think.

**Comment by dkerr64** [comment]
> Looking nice. I think we can combine the separate regexes into a single variable without losing readability. I pushed a commit to your branch that does just that; please take a look and tell me what you think.

Yes combining them makes sense.  I see you also merged in all the group definitions that I had separated out into their own `qr/ /`.  I did that because they are all required again when it comes to the GUA version of the regex's and it is better to have them defined only once.  But I am keen to get this merged in so if you want to go with this now that is fine, I will update the other pull request for the GUA version which can either put them back into their own `qr/ /` or duplicate them.

I have a bunch of IPv6 stuff waiting in the wings that need this (and the GUA version) merged before I can offer those up to you.

Thanks.

**Comment by rhansen** [comment]
I pushed a couple of commits to your branch, mostly to improve the tests. Please take a look.

> I did that because they are all required again when it comes to the GUA version of the regex's

Ah, right, I forgot about that. You can extract them to a separate variable in PR #197.

**Comment by dkerr64** [comment]
> Ah, right, I forgot about that. You can extract them to a separate variable in PR #197.

yes I'll take care of it then.

**Comment by rhansen** [comment]
Merged. Thank you!
